relay-server/src/services/processor.rs

```diff
@@ -801,11 +785,41 @@ struct EventMetricsExtracted(bool);
 #[derive(Debug, Copy, Clone)]
 struct SpansExtracted(bool);
 
-/// The view out of the [`ProcessEnvelopeState`] after processing.
+/// The result of the envelope processing containing the processed envelope along with the partial
+/// result.
 #[derive(Debug)]
-struct ProcessingStateResult {
+struct ProcessingResult {
     managed_envelope: TypedEnvelope<Processed>,
-    extracted_metrics: ExtractedMetrics,
+    partial: ProcessingPartialResult,
+}
+
+impl ProcessingResult {
+    /// Creates a [`ProcessingResult`] with no metrics.
+    fn no_metrics(managed_envelope: TypedEnvelope<Processed>) -> Self {
+        Self {
+            managed_envelope,
+            partial: ProcessingPartialResult::NoMetrics,
+        }
+    }
+
+    /// Returns the components of the [`ProcessingResult`].
+    fn into_inner(self) -> (TypedEnvelope<Processed>, Option<ExtractedMetrics>) {
+        match self.partial {
+            ProcessingPartialResult::NoMetrics => (self.managed_envelope, None),
+            ProcessingPartialResult::WithMetrics { extracted_metrics } => {
+                (self.managed_envelope, Some(extracted_metrics.metrics))
+            }
+        }
+    }
+}
+
+/// The partial result of the envelope processing.
+#[derive(Debug)]
+enum ProcessingPartialResult {
```

Comment:
Created this partial result to be more clear about the fact that this result is not the full one because the envelope is added afterwards.

Reply:
Can we have `ProcessingExtractedMetrics` on the `ProcessingResult` instead? That would make it simpler to understand IMO, with a PartialResult type the reader has to follow one additional hop.